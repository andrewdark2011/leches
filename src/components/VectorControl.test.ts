import { mount } from '@vue/test-utils'
import { TresLeches, useControls } from '/@/'
import { it, expect, describe } from 'vitest'
import { defineComponent, nextTick } from 'vue'

class Vector3 {
  constructor( x = 0, y = 0, z = 0 ) {

		Vector3.prototype.isVector3 = true;

		this.x = x;
		this.y = y;
		this.z = z;

	}

  set( x, y, z ) {

		if ( z === undefined ) z = this.z; // sprite.scale.set(x,y)

		this.x = x;
		this.y = y;
		this.z = z;

		return this;

	}

  *[ Symbol.iterator ]() {

		yield this.x;
		yield this.y;
		yield this.z;

	}
}
describe('Vector Control', async () => {
  it('should render 3 numeric inputs', async ()=> {
    const component = defineComponent({
      template: `
        <TresLeches />
      `,
      components: {
        TresLeches
      },
      setup() {
        const { position } = useControls({ position: new Vector3(0, 2, 4) })
        return {
          position
        }
      }
    })
    const wrapper = mount(component, {
      components: {
        TresLeches
      },
    })
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.findAll('input[type="number"]').length).toBe(3)
  })
  it('should render a number control with a label', async ()=> {
    const component = defineComponent({
      template: `
        <TresLeches />
      `,
      components: {
        TresLeches
      },
      setup() {
        const { position } = useControls({ position: new Vector3(0, 2, 4) })
        return {
          position
        }
      }
    })

    const wrapper = mount(component, {
      components: {
        TresLeches
      },
    })
    expect(wrapper.find('label').text()).toBe('position')
  })
  it('should change the value of the control when any of the input changes', async ()=> {
    const component = defineComponent({
      template: `
        <TresLeches />
      `,
      components: {
        TresLeches
      },
      setup() {
        const { position } = useControls({ position: new Vector3(0, 2, 4) })
        return {
          position
        }
      }
    })

    const wrapper = mount(component, {
      components: {
        TresLeches
      },
    })
    const inputs = wrapper.findAll('input[type="number"]')

    await inputs[0].setValue(20)
    await inputs[1].setValue(-20)
    await inputs[2].setValue(-1)
    await nextTick();
    console.log(wrapper.vm.position)
    expect(wrapper.vm.position.x).toBe(20)
    expect(wrapper.vm.position.y).toBe(-20)
    expect(wrapper.vm.position.z).toBe(-1)
  })
  // TODO: mouseDown
})