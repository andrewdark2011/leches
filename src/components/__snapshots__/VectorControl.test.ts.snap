// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Vector Control > should render multiple numeric inputs 1`] = `
"<div style=\\"left: 704px; top: 10px;\\" class=\\"absolute select-none z-24 w-280px font-sans text-xs\\">
  <div tabindex=\\"0\\" class=\\"bg-white shadow-xl rounded border-4 border-solid border-black\\">
    <header class=\\"relative cursor-grabbing p-4 flex justify-between text-gray-200 relative\\"><i class=\\"h-4 w-4 p-1.5 flex items-center line-height-0 rounded-full bg-gray-100 text-xs\\">🍰</i>
      <div><i class=\\"i-ic-baseline-drag-indicator\\"></i><i class=\\"i-ic-baseline-drag-indicator\\"></i><i class=\\"i-ic-baseline-drag-indicator\\"></i></div>
      <div></div>
    </header>
    <!--v-if-->
    <div class=\\"flex px-4 justify-between gap-1 items-center mb-2\\"><label for=\\"position\\" title=\\"position\\" class=\\"text-gray-500 w-1/3 truncate\\">position</label>
      <div class=\\"relative w-2/3 flex justify-between gap-0.5\\">
        <div class=\\"flex items-center bg-gray-100 rounded w-1/3 w-1/2\\"><span class=\\"font-bold px-1 py-1 text-0.65rem text-gray-300\\">x</span><input id=\\"position-x\\" type=\\"number\\" step=\\"1\\" class=\\"w-full px-0 p-1 text-right text-0.65rem text-gray-400 bg-transparent focus:border-gray-200 outline-none border-none font-sans appearence-none\\"></div>
        <div class=\\"flex items-center bg-gray-100 rounded w-1/3 w-1/2\\"><span class=\\"font-bold px-1 py-1 text-0.65rem text-gray-300\\">y</span><input id=\\"position-y\\" type=\\"number\\" step=\\"1\\" class=\\"w-full px-0 p-1 text-right text-0.65rem text-gray-400 bg-transparent focus:border-gray-200 outline-none border-none font-sans appearence-none\\"></div>
        <div class=\\"flex items-center bg-gray-100 rounded w-1/3 w-1/2\\"><span class=\\"font-bold px-1 py-1 text-0.65rem text-gray-300\\">z</span><input id=\\"position-z\\" type=\\"number\\" step=\\"1\\" class=\\"w-full px-0 p-1 text-right text-0.65rem text-gray-400 bg-transparent focus:border-gray-200 outline-none border-none font-sans appearence-none\\"></div>
      </div>
    </div>
  </div>
</div>"
`;
